pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java

```diff
@@ -11,41 +10,76 @@
 import org.jaxen.JaxenException;
 import org.w3c.dom.Document;
 
+import net.sourceforge.pmd.annotation.InternalApi;
 import net.sourceforge.pmd.lang.ast.xpath.Attribute;
 import net.sourceforge.pmd.lang.dfa.DataFlowNode;
 
 /**
- * All AST nodes must implement this interface. It provides basic
- * machinery for constructing the parent and child relationships
- * between nodes.
+ * Root interface for all AST nodes. This interface provides only the API
+ * shared by all AST implementations in PMD language modules. This includes for now:
+ * <ul>
+ * <li>Tree traversal methods, like {@link #getParent()} and {@link #getFirstChildOfType(Class)}
+ * <li>The API used to describe nodes in a form understandable by XPath expressions:
+ * {@link #getXPathNodeName()},  {@link #getXPathAttributesIterator()}
+ * <li>Location metadata: eg {@link #getBeginLine()}, {@link #getBeginColumn()}
+ * </ul>
+ *
+ * <p>It is assumed in many places that the {@link #getChild(int)} and {@link #getParent()}
+ * method return a node of the same type than this node. For example,
+ * no JSP node should have a Java node as its child. Embedding nodes from
+ * different languages will not be done via these methods, and conforming
+ * implementations should ensure, that every node returned by these methods
+ * are indeed of the same type. Possibly, a type parameter will be added to
+ * the Node interface in 7.0.0 to enforce it at compile-time.
+ *
+ * <p>Note: I'm pretty sure Apex doesn't respect this invariant with its comment nodes.
```

Comment:
I think, this comment is not helpful here :smile: 

What's the criteria to keep this invariant? Do all nodes need to share the same (abstract) base class? Looking at the current Apex comment implementation, I think, it works already that way: ASTFormalComment subclasses AbstractApexNodeBase. The only difference is, that these nodes do not originate from the parser, but we added them later into the AST manually (in ApexTreeBuilder).

Do we need to change there anything? Then I'd like to create a issue for that...

Reply:
Uh right probably not the best place :smile: 

> What's the criteria to keep this invariant? Do all nodes need to share the same (abstract) base class?

I mean, share a published interface. We have `JavaNode`, `ModelicaNode`, etc. Apex uses an abstract class which should be internal. 

> The only difference is, that these nodes do not originate from the parser, but we added them later into the AST manually (in ApexTreeBuilder).

This is why the difference should not be published API, it's an implementation detail. Apex comments should implement the regular ApexNode.